Sources/SmithyReadWrite/SmithyReader.swift

```diff
@@ -0,0 +1,203 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import struct Foundation.Data
+import struct Foundation.Date
+import enum SmithyTimestamps.TimestampFormat
+
```

Comment:
This protocol is what both readers conform to (XML & JSON) so that they can share code for reading closures, etc.

`readIfPresent()` methods are required in concrete conforming types to read an optional of each basic member type.  Non-optional `read()` methods are extended directly onto this protocol for reading non-optional values.